internal/webhooks/cluster.go

```diff
@@ -485,3 +448,103 @@ func validateCIDRBlocks(fldPath *field.Path, cidrs []string) field.ErrorList {
+			if md.Variables == nil || len(md.Variables.Overrides) == 0 {
+				continue
+			}
+			defaultedVariables, errs := variables.DefaultMachineDeploymentVariables(md.Variables.Overrides, clusterClassStatusVariablesToVariables(clusterClass.Status.Variables),
+				field.NewPath("spec", "topology", "workers", "machineDeployments").Index(i).Child("variables", "overrides"))
+			if len(errs) > 0 {
+				allErrs = append(allErrs, errs...)
+			} else {
+				md.Variables.Overrides = defaultedVariables
+			}
+		}
+	}
+	return allErrs
+}
+
+// ValidateClusterForClusterClass uses information in the ClusterClass to validate the Cluster.
+func ValidateClusterForClusterClass(cluster *clusterv1.Cluster, clusterClass *clusterv1.ClusterClass, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if cluster == nil {
+		return field.ErrorList{field.InternalError(field.NewPath(""), errors.New("Cluster can not be nil"))}
+	}
+	if clusterClass == nil {
+		return field.ErrorList{field.InternalError(field.NewPath(""), errors.New("ClusterClass can not be nil"))}
+	}
+	allErrs = append(allErrs, check.MachineDeploymentTopologiesAreValidAndDefinedInClusterClass(cluster, clusterClass)...)
+
+	// validate the MachineHealthChecks defined in the cluster topology
+	allErrs = append(allErrs, validateMachineHealthChecks(cluster, clusterClass)...)
+
+	// Check if the variables defined in the ClusterClass are valid.
+	allErrs = append(allErrs, variables.ValidateClusterVariables(cluster.Spec.Topology.Variables, clusterClassStatusVariablesToVariables(clusterClass.Status.Variables),
+		fldPath.Child("variables"))...)
+
+	if cluster.Spec.Topology.Workers != nil {
+		for i, md := range cluster.Spec.Topology.Workers.MachineDeployments {
+			// Continue if there are no variable overrides.
+			if md.Variables == nil || len(md.Variables.Overrides) == 0 {
+				continue
+			}
+			allErrs = append(allErrs, variables.ValidateMachineDeploymentVariables(md.Variables.Overrides, clusterClassStatusVariablesToVariables(clusterClass.Status.Variables),
+				fldPath.Child("workers", "machineDeployments").Index(i).Child("variables", "overrides"))...)
+		}
+	}
+
+	return allErrs
+}
+
+// pollClusterClassForCluster will retry getting the ClusterClass referenced in the Cluster for two seconds.
+func (webhook *Cluster) pollClusterClassForCluster(ctx context.Context, cluster *clusterv1.Cluster) (*clusterv1.ClusterClass, error) {
+	clusterClass := &clusterv1.ClusterClass{}
+	var clusterClassGetErr error
+	_ = util.PollImmediate(200*time.Millisecond, 2*time.Second, func() (bool, error) {
+		if clusterClassGetErr = webhook.Client.Get(ctx, client.ObjectKey{Namespace: cluster.Namespace, Name: cluster.Spec.Topology.Class}, clusterClass); clusterClassGetErr != nil {
+			return false, nil //nolint:nilerr
+		}
+		// Return an error if a ClusterClass has variables defined, but no variables in its ClusterClass.Status.
+		if clusterClass.Status.ObservedGeneration != clusterClass.GetGeneration() {
```

Comment:
This case is a little bit complex. The intention is to make it so a Cluster's variables are defaulted and validated if the ClusterClass exists and its variables have been correctly reconciled but:
1) A generation check introduces new dependencies i.e. with all of the templates used in the ClusterClass
2) A `spec.Variables` check ignores any variables that might come from Runtime Extensions
3) A runtimeExtension can reasonably return no variables, which means we can't use `len(Status.Variables)== 0` to assert that variables have not been correctly updated.

I'm leaning toward ignoring whether or not there are variables from Runtime SDK here - only validating if some variables are defined in `spec.Variables` as Runtime Variables can reasonably change after the webhook anyway.

WDYT? @sbueringer @fabriziopandini @ykakarap

Reply:
I think the cleanest way to implement this is probably with a new specific condition for reconciled variables. This has the added benefit of surfacing the errors to users.

Reply:
Discussed this with @sbueringer decided the best way forward is to implement a condition here and also check for the observedGeneration to make sure that both RuntimeExtension changes and Clusterclass spec changes are caught here.

The outstanding issue is to fix current tests so that the generation check allows them to pass - i.e. by adding real templates to tests etc. and to allow this check to work with DryRun, either by adding a Clusterclass reconcile to dry run or adding the condition at the start of dry run.